src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -467,13 +467,34 @@ _public_ int sd_pid_notify_with_fds(
         if (!e)
                 return 0;
 
-        r = sockaddr_un_set_path(&sockaddr.un, e);
+        /* Allow AF_UNIX and AF_VSOCK, reject the rest. */
+        r = socket_local_address_parse(&address, e);
         if (r < 0)
                 goto finish;
-        msghdr.msg_namelen = r;
+        msghdr.msg_namelen = address.size;
+
+        /* If we didn't get an address, talk to the hypervisor by default, as that's the purpose of this functionality. */
+        if (address.sockaddr.vm.svm_family == AF_VSOCK && address.sockaddr.vm.svm_cid == VMADDR_CID_ANY)
+                address.sockaddr.vm.svm_cid = VMADDR_CID_HYPERVISOR;
```

Comment:
hmm, shouldn't this be VMADDR_CID_HOST?

(no idea what the precise difference is supposed to be?)

Reply:
hmm, i wonder how much sense this automatic fallback makes? maybe just return error here?

Reply:
_HYPVERVISOR allows to talk only the hypervisor process, while _HOST allows talking to any process that will listen on the host (I don't actually know how vhost drivers implement this, going by the definition alone). Defaulting to _HOST would mean anything could start listening to state changes in the guests, and might not be desirable (unless it was explicitly configured). I figure talking to the hypervisor is always safe though, as that has privileges over guests by definition. The CID is most often documented as optional with the "right thing" being done if omitted, so this seems to me to be the path of least surprise for consumers.

Reply:
it's not clear to me that the "right thing" is to talk to the hypervisor. In my understanding in a libvirt/qemu world, it would be libvrit that gets the messages, not qemu. And i'd assume that _HYPERVISOR ends up only in qemu, not in libvirt. But not sure how this is supposed to work and if i grokked this properly.

But something tells me we should maybe not fall back to a default here if things aren't so clear cut.

Reply:
Sure but libvirt also configures the VM, so it can set _HOST. The point is that many documents talking about vsock say the CID is optional, so for consistency we should do the same. And _HOST does not seem like a safe default.

Reply:
dunno, still not convinced that doing _HYPERVISOR is really that great a cid to default to here..

I mean, I can see that _HOST is not a safe default. but _HYPERVISOR might be "safe" but I think in almost all cases *wrong*...

so I'd really just fail parsing here.

(but also, i don't care too much. if you insist)

Reply:
I insist :-) We can always change it later when the first consumer shows up, it's all academic right now.

Reply:
nah, you can't change it later that would be a compat break.

you can't just make something we accept as valid now meaning one thing to mean something totally different.

if we'd generate an error here now, then we'd be on the safe side otoh: we have all options to make it mean whatever we want later.